File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/ObjC/FileService/FileService/BrowseViewController.m
#import "BrowseViewController.h"
#import "ImagePreviewViewController.h"
#import "FileObject.h"
#import <Backendless-Swift.h>

@interface BrowseViewController () {
    NSArray *mainData;
    DataStoreFactory *filesDataStore;
}
@end

@implementation BrowseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    filesDataStore = [Backendless.shared.data of:[FileObject class]];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self getAllEntitysAsync];
}

- (void)getAllEntitysAsync {
    [filesDataStore findWithResponseHandler:^(NSArray *photos) {
        if (photos.count == 0) {
            [[[UIAlertView alloc] initWithTitle:@"No files found" message:@"Please take some photos before you can browse them" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil] show];
            [self.navigationController popToRootViewControllerAnimated:YES];
        }
        else {
            self->mainData = photos;
            [self.mainTableView reloadData];
        }
    } errorHandler:^(Fault *fault) {
        [[[UIAlertView alloc] initWithTitle:@"Error" message:fault.message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
        [self.navigationController popToRootViewControllerAnimated:YES];
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return (mainData.count / 4) + ((mainData.count % 4) ? 1 : 0);
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    for (int i = 1; i <= 4; i++) {
        if (indexPath.row * 4 + i > mainData.count)
            break;
        UIButton *button = (UIButton *)[cell viewWithTag:i];
        NSString *str = (NSString *)[[mainData objectAtIndex:indexPath.row * 4 + i - 1] path];
        NSURL *url = [NSURL URLWithString:str];
        button.hidden = NO;
        button.enabled = NO;
        [NSURLConnection sendAsynchronousRequest:[NSURLRequest requestWithURL:url] queue:[NSOperationQueue mainQueue] completionHandler:^(NSURLResponse *response, NSData *data, NSError *error) {
            NSHTTPURLResponse *responseUrl = [(NSHTTPURLResponse *)response copy];
            NSInteger statusCode = [responseUrl  statusCode];
            if (statusCode == 200) {
                [button setImage:[UIImage imageWithData:data] forState:UIControlStateNormal];
                button.enabled = YES;
            }
        }];
    }
    return cell;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSIndexPath *indexPath = [self.mainTableView indexPathForCell:(UITableViewCell *)[[sender superview] superview]];
    ImagePreviewViewController *imagePreview = (ImagePreviewViewController *)[segue destinationViewController];
    [imagePreview setMainImage:[(UIButton *)sender imageForState:UIControlStateNormal]];
    [imagePreview setIsUpload:YES];
    [imagePreview setFile:[mainData objectAtIndex:indexPath.row * 4 + [sender tag] - 1]];
}

- (IBAction)pressedRemoveAll:(id)sender {
    [Backendless.shared.file removeWithPath:@"img" responseHandler:^{
        for (FileObject *file in self->mainData) {
            [self->filesDataStore removeWithEntity:file responseHandler:^(NSInteger removed) {
                [[[UIAlertView alloc] initWithTitle:@"Files removed" message:@"All files have been removed" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
            } errorHandler:^(Fault *fault) {
                [[[UIAlertView alloc] initWithTitle:@"Error" message:fault.message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
            }];
        }
    } errorHandler:^(Fault *fault) {
        [[[UIAlertView alloc] initWithTitle:@"Error" message:fault.message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil] show];
    }];
    [self.navigationController popToRootViewControllerAnimated:YES];
}

@end

File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/Swift/FileService/FileService/BrowseViewController.swift

import UIKit
import Backendless

class BrowseViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var mainTableView: UITableView!
    
    var mainData: [FileObject]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        getAllEntitiesAsync()
    }
    
    func getAllEntitiesAsync() {
        Backendless.shared.data.of(FileObject.self).find(responseHandler: { photos in
            if photos.count == 0 {
                UIAlertView.init(title: "No files found", message: "Please take some photos before you can browse them", delegate: nil, cancelButtonTitle: "OK").show()
                self.navigationController?.popToRootViewController(animated: true)
            }
            else {
                self.mainData = photos as? [FileObject]
                self.mainTableView.reloadData()
            }
        }, errorHandler: { fault in
            UIAlertView.init(title: "Error", message: fault.message ?? "", delegate: nil, cancelButtonTitle: "OK").show()
            self.navigationController?.popToRootViewController(animated: true)
        })
    }
    
    @IBAction func removeAll(_ sender: Any) {
        Backendless.shared.file.remove(path: "img", responseHandler: {
            for file in self.mainData! {
                Backendless.shared.data.of(FileObject.self).remove(entity: file, responseHandler: { removed in
                    UIAlertView.init(title: "Files removed", message: "All files have been removed", delegate: nil, cancelButtonTitle: "OK").show()
                    self.navigationController?.popToRootViewController(animated: true)
                }, errorHandler: { fault in
                    UIAlertView.init(title: "Error", message: fault.message ?? "", delegate: nil, cancelButtonTitle: "OK").show()
                    self.navigationController?.popToRootViewController(animated: true)
                })
            }
        }, errorHandler: { fault in
            UIAlertView.init(title: "Error", message: fault.message ?? "", delegate: nil, cancelButtonTitle: "OK").show()
            self.navigationController?.popToRootViewController(animated: true)
        })
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if mainData != nil {
            if (mainData!.count % 4 == 0) {
                return (mainData!.count / 4)
            }
            else {
                return (mainData!.count / 4 + 1)
            }
        }
        else {
            return 0
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "Cell"
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        for i in 1...4 {
            if (indexPath.row * 4 + i > (mainData?.count)!) {
                break
            }
            let button = cell?.viewWithTag(i) as! UIButton
            if let str = mainData?[indexPath.row * 4 + i - 1].path {
                let url = NSURL(string: str)
                button.isHidden = false
                button.isEnabled = false
                NSURLConnection.sendAsynchronousRequest(NSURLRequest.init(url: url! as URL) as URLRequest, queue: OperationQueue.main, completionHandler: { (response: URLResponse?, data: Data?, error: Error?) -> () in
                    let responseUrl = response?.copy() as! HTTPURLResponse
                    let statusCode = responseUrl.statusCode
                    if (statusCode == 200) {
                        button.setImage(UIImage.init(data: data!), for: .normal)
                        button.isEnabled = true
                    }
                })
            }
        }
        return cell!
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let imagePreview = segue.destination as? ImagePreviewViewController {
            imagePreview.mainImage = (sender as! UIButton).image(for: .normal)
            imagePreview.isUpload = true
            let cell = (sender as! UIButton).superview?.superview as! UITableViewCell
            let row = mainTableView.indexPath(for: cell)!.row
            imagePreview.file = mainData?[row * 4 + ((sender as AnyObject).tag)! - 1]
        }
    }
}
